smart/model.py

```diff
@@ -998,6 +1005,71 @@ def _init_4_7_set_initial_conditions(self):
+            if parameter.type == ParameterType.from_xdmf:
+                # load the time vec from xdmf file
+                xdmf_file = open(parameter.xdmf_file, "r")
+                xdmf_string = xdmf_file.read()
+                found_pattern = re.findall(r"Time Value=\"?[^\s]+", xdmf_string)
+                tVec = []
+                for i in range(len(found_pattern)):
+                    tVec.append(float(found_pattern[i][12:-1]))
+                parameter.tVec = np.array(tVec)
+
+                # define function space
+                if parameter.compartment not in self.cc.keys:
+                    raise ValueError(
+                        f"Compartment name {parameter.compartment} for parameter"
+                        f"{parameter.name} does not match a known compartment"
+                    )
+                V_cur = d.FunctionSpace(self.cc[parameter.compartment].dolfin_mesh, "P", 1)
+                parameter.dolfin_function = d.Function(V_cur)
+
+                cur_file = d.HDF5File(self.parent_mesh.mpi_comm, parameter.h5_file, "r")
+                # Find index associated with the starting time
+                if np.any(np.isclose(tVec, float(self.t))):
+                    vec_new = d.Vector()
+                    idx1 = np.nonzero(np.isclose(tVec, float(self.t)))[0][0]
+                    cur_file.read(vec_new, f"VisualisationVector/{idx1}", True)
+                    cur_file.close()
+                elif self.t > tVec[-1]:  # then starting after final time in the xdmf file
+                    logger.warning(
+                        f"File {str(parameter.h5_file)} ends before current time {self.t}"
+                        "Using final time point in file instead."
+                    )
+                    vec_new = d.Vector()
+                    idx1 = len(tVec) - 1
+                    cur_file.read(vec_new, f"VisualisationVector/{idx1}", True)
+                    cur_file.close()
+                elif self.t < tVec[0]:  # then starting before initial time in xdmf file
+                    logger.warning(
+                        f"File {str(parameter.h5_file)} starts after current time {self.t}"
+                        "Using initial time point in file instead."
+                    )
+                    vec_new = d.Vector()
+                    idx1 = 0
+                    cur_file.read(vec_new, f"VisualisationVector/{idx1}", True)
+                    cur_file.close()
+                else:  # then in between two times in the xdmf file
+                    vec1 = d.Vector()
+                    vec2 = d.Vector()
+                    idx1 = np.nonzero(tVec < float(self.t))[0][-1]
+                    idx2 = np.nonzero(tVec > float(self.t))[0][0]
+                    cur_file.read(vec1, f"VisualisationVector/{idx1}", True)
+                    cur_file.read(vec2, f"VisualisationVector/{idx2}", True)
+                    vec_new = (vec1 + vec2) / 2
+                vec = parameter.dolfin_function.vector()
```

Comment:
Did your forget to close the file here? Might just add a single `cur_file.close` at the end or wrapping it in a context manager. 

Our just make a function
```python
def load_vector(h5_file, comm, t, tVec):
     with d.HF5File(comm, h5file, "r"):
        # All conditional logic
     return vec_new
```

Reply:
I like how your previous comment predicted this would happen right before it did! :) I guess it has always hit another case after hitting this one and successfully closed the hdf5 file, this is an excellent catch

Reply:
Function is written and things are running smoothly